reframe/core/config.py

```diff
@@ -217,3 +220,121 @@ def create_env(system, partition, name):
+        if 'partitions' in sys_specs:
+            sys_dict['partitions'] = []
+            for pname, p in sys_specs['partitions'].items():
+                new_p = {'name': pname}
+                new_p.update(p)
+                if p['scheduler'] == 'nativeslurm':
+                    new_p['scheduler'] = 'slurm'
+                    new_p['launcher'] = 'srun'
+                elif p['scheduler'] == 'local':
+                    new_p['scheduler'] = 'local'
+                    new_p['launcher'] = 'local'
+                else:
+                    sched = p['scheduler'].split('+')[0]
+                    launch = p['scheduler'].split('+')[1]
+                    new_p['scheduler'] = sched
+                    new_p['launcher'] = launch
+
+                # Make resources dictionary into a list
+                if 'resources' in p:
+                    new_p['resources'] = []
+                    for rname, r in p['resources'].items():
+                        new_r = {'name': rname, 'options': r}
+                        new_p['resources'].append(new_r)
+
+                # Make variables dictionary into a list of lists
+                if 'variables' in p:
+                    new_p['variables'] = []
+                    for vname, v in p['variables'].items():
+                        new_p['variables'].append([vname, v])
+
+                sys_dict['partitions'].append(new_p)
+
+        converted['systems'].append(sys_dict)
+
+    old_environs = old_config.site_configuration['environments'].items()
+    for env_target, env_entries in old_environs:
+        for ename, e in env_entries.items():
+            new_env = {'name': ename}
+            if (env_target != '*'):
+                new_env['target_systems'] = [env_target]
+
+            new_env.update(e)
+            converted['environments'].append(new_env)
+
+    if 'modes' in old_config.site_configuration:
+        converted['modes'] = []
+        old_modes = old_config.site_configuration['modes'].items()
+        for target_mode, mode_entries in old_modes:
+            for mname, m in mode_entries.items():
+                new_mode = {'name': mname, 'options': m}
+                if target_mode != '*':
+                    new_mode['target_systems'] = [target_mode]
+
+                converted['modes'].append(new_mode)
+
+    def update_config(log_name, original_log):
+        new_handlers = []
+        for h in original_log['handlers']:
+            new_h = h
+            new_h['level'] = h['level'].lower()
+            new_handlers.append(new_h)
+
+        converted[log_name].append(
+            {
+                'level': original_log['level'].lower(),
+                'handlers': new_handlers
+            }
+        )
+
+    update_config('logging', old_config.logging_config)
+    update_config('perf_logging', old_config.perf_logging_config)
+
+    if (hasattr(old_config, 'checks_path') or
+        hasattr(old_config, 'checks_path_recurse')):
```

Comment:
To avoid this check here, I would put an empty entry for `general` in the new dictionary in any case and just fill it in as you have it below. At the end I would check if it's still empty and if it is, I'd delete it.